argus/scenarios/service_mock.py

```diff
@@ -88,11 +89,47 @@ def stop_me(self):  # pylint: disable=no-self-use
         cherrypy.engine.exit()
 
 
-class EC2MetadataServiceApp(BaseServiceApp):
-    pass
+class MetadataServiceAppMixin(object):
+    """Common metadata resources."""
+
+    def instance_id(self, *_):
+        return self.scenario._server['id']
+
+    def local_hostname(self, *_):
+        return self.scenario.instance_server()['name'][:15].lower()
+
+    def public_keys(self, *_):
+        return self.scenario.public_key().strip()
 
 
-class CloudstackMetadataServiceApp(BaseServiceApp):
+class EC2MetadataServiceApp(BaseServiceApp, MetadataServiceAppMixin):
+
+    def __init__(self, *args, **kwargs):
+        super(EC2MetadataServiceApp, self).__init__(*args, **kwargs)
+        self._keydict = None
+
+    def _store_keydict(self):
+        if self._keydict:
+            return
+        keys = super(EC2MetadataServiceApp, self).public_keys().splitlines()
+        self._keydict = dict(enumerate(keys))
+
+    @cherrypy.expose
+    def default(self, *args, **_):
+        operation, remain = args[0], args[1:]
+        return self._dispatch_method(operation)(remain)
+
+    def public_keys(self, remain):
+        self._store_keydict()
```

Comment:
Why not a property then? Having to call a function to save state in order to continue operating with that state  is error prone.